### PR TITLE
Pimple PSR-11 implementation

### PR DESCRIPTION
**Pull request by paradajozsef**
Back in 2014 there was a proposal for container-interop: #162, which was rejected. 

PSR-11 is now [in draft status](http://www.php-fig.org/psr/#draft) and planty of [projects implementing it](https://github.com/container-interop/container-interop#projects-implementing-containerinterface).

I don't know if draft status is enough for Pimple to implement it, I'd like to know it. That is the goal of this PR.

What is your opinion?

### CODE REVIEW

**Comment by Sam-Burns** [comment]
:+1:

**Comment by Sam-Burns** [comment]
Some of your builds will start working if you can run `composer install` in the `.travis.yml` file, instead of just dumping the autoloader.  Not sure what needs to be done about the extension builds though.

**Comment by kdubuc** [comment]
👍

**Comment by kdubuc** [comment]
If Pimple becomes PSR11 compliant, it will be incompatible with Silex, because Silex extends Pimple, and there will be a conflict with the get function  https://github.com/silexphp/Silex/blob/1.3/src/Silex/Application.php#L227 vs. https://github.com/container-interop/container-interop/blob/master/src/Interop/Container/ContainerInterface.php#L26 😞

**Comment by paradajozsef** [comment]
I apologize to everyone for late answer. (I've been lack of internet at home lately because of moving to another place.)

Thanks for the reviews!

@Sam-Burns Thanks, it made the no extension builds green. :)

@kdubuc You are right, great catch! I guess it can't be accepted until a new major version is out from Pimple and Silex too.

**Comment by hkdobrev** [comment]
@paradajozsef BTW PSR-11 now has released the Composer package with the interfaces and exception classes under the PSR namespace.

File: src/Pimple/Container.php [hunk 1]
```diff
@@ -58,6 +63,22 @@ public function __construct(array $values = array())
     }
 
     /**
+     * {@inheritdoc}
+     */
+    public function get($id)
```
Comment (GromNaN) [inline src/Pimple/Container.php]: This conflict with [`Silex\Application::get()`](https://github.com/silexphp/Silex/blob/v2.0.3/src/Silex/Application.php#L143). That's a huge backward compatibility break for both Pimple and Silex.

**Comment by GromNaN** [comment]
Alternatively, I've created a library that bridge Pimple Container to the ContainerInterface.
https://github.com/GromNaN/pimple-interop

**Comment by hkdobrev** [comment]
@GromNaN There are already two packages supporting the Container Interop project: 

- https://github.com/moufmouf/pimple-interop
- https://github.com/Sam-Burns/pimple3-containerinterop

Yours is probably the first which uses the new packages under the `psr/` vendor name on Packagist.

However all these projects while supporting container interoperability prevent Pimple to be used with Silex because Silex extends the container and has a `get` method for `GET` routes. This is something which would be great to be resolved by the official Pimple which is made to work with Silex.

**Comment by GromNaN** [comment]
Found also this one: https://github.com/snapshotpl/pimple-interop

Before trying to make Pimple compatible with [PSR-11](https://github.com/php-fig/fig-standards/blob/master/proposed/container.md), we need to think to actual use cases. The purpose of the whole `ContainerInterface` it to "standardize how frameworks and libraries make use of a container to obtain objects and parameters". That way a class can accept any container implementation. Due to the conflict on the `get` method, a `Silex\Application` can't implement the `ContainerInterface`. This is not necessary at all!

Unlike other libraries, mine use the adapter pattern by delegation over inheritance. I don't try to make Pimple PSR-11 compliant: I expose Pimple thrown the `ContainerInterface`. And this is compatible with Silex.

```php
class FooService {
    function __construct(ContainerInterface $container) {}
}

$app = new Silex\Application();
$app->register(new GromNaN\Pimple\PimpleContainerProvider());
$app['service'] = function ($app) {
    return new FooService($app['container']);
};
```

**Comment by fabpot** [comment]
Closing as PSR-11 is now fully supported.

**Pull request closed without merging**